File: ETSMobile/Controllers/Base/ETSAuthenticationViewController.m
//
//  ETSAuthenticationViewController.m
//  ETSMobile
//
//  Created by Jean-Philippe Martin on 2013-11-03.
//  Copyright (c) 2013 ApplETS. All rights reserved.
//

#import "ETSAuthenticationViewController.h"
#import "KeychainItemWrapper.h"
#import "AppDelegate.h"
#import "MSDynamicsDrawerViewController.h"

NSString * const kKeychainId = @"ApplETS";

@interface ETSAuthenticationViewController ()

@end

@implementation ETSAuthenticationViewController

- (IBAction)panLeftMenu:(id)sender
{
    MSDynamicsDrawerViewController *dynamicsDrawerViewController = ((AppDelegate *)[[UIApplication sharedApplication] delegate]).dynamicsDrawerViewController;
    [dynamicsDrawerViewController setPaneState:MSDynamicsDrawerPaneStateOpen animated:YES allowUserInterruption:YES completion:^{}];
}

+ (ETSSynchronizationResponse) validateJSONResponse:(NSDictionary *)response
{
    id apiError = [response valueForKeyPath:@"d.erreur"];
    id requestError = [response valueForKeyPath:@"Message"];
    
    if ([apiError isKindOfClass:[NSString class]] && [apiError isEqualToString:@"Code d'accès ou mot de passe invalide"]) {
        return ETSSynchronizationResponseAuthenticationError;
    } else if ([apiError isKindOfClass:[NSString class]] && [apiError length] == 0) {
        return ETSSynchronizationResponseValid;
    } else if ([apiError isKindOfClass:[NSString class]] && [apiError length] > 0) {
        return ETSSynchronizationResponseUnknownError;
    } else if ([requestError isKindOfClass:[NSString class]] && [requestError isEqualToString:@"There was an error processing the request."]) {
        return ETSSynchronizationResponseAuthenticationError;
    }
    
    return ETSSynchronizationResponseValid;
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    [self.navigationController setNavigationBarHidden:[[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPhone animated:animated];
    
    if ([[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPad) {
        self.menuButton.hidden = YES;
        self.cancelButton.hidden = NO;
    } else {
        self.cancelButton.hidden = YES;
    }
    
    MSDynamicsDrawerViewController *dynamicsDrawerViewController = ((AppDelegate *)[[UIApplication sharedApplication] delegate]).dynamicsDrawerViewController;
    [dynamicsDrawerViewController setPaneDragRevealEnabled:YES forDirection:MSDynamicsDrawerDirectionLeft];
}

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    [self.usernameTextField becomeFirstResponder];
}

- (BOOL)textFieldShouldReturn:(UITextField *)textField
{
    if (textField == self.usernameTextField) {
        [self.passwordTextField becomeFirstResponder];
    }
    else {
        if ([self.usernameTextField.text length] == 0) [self.usernameTextField becomeFirstResponder];
        else {
            [self authenticate:nil];
        }
    }
    return YES;
}

- (IBAction)authenticate:(id)sender
{
    [self.usernameTextField resignFirstResponder];
    [self.passwordTextField resignFirstResponder];
    
    KeychainItemWrapper *keychainItem = [[KeychainItemWrapper alloc] initWithIdentifier:@"ApplETS" accessGroup:nil];
    [keychainItem setObject:self.passwordTextField.text forKey:(__bridge id)(kSecValueData)];
    [keychainItem setObject:self.usernameTextField.text forKey:(__bridge id)(kSecAttrAccount)];
    
    [self.delegate controllerDidAuthenticate:self];
}

- (IBAction)cancel:(id)sender
{
    [self dismissViewControllerAnimated:YES completion:nil];
}

+ (NSString *)passwordInKeychain
{
    KeychainItemWrapper *keychainItem = [[KeychainItemWrapper alloc] initWithIdentifier:kKeychainId accessGroup:nil];
    NSString *password = [keychainItem objectForKey:(__bridge id)(kSecValueData)];
    if ([password length] <= 0) password = nil;
    return password;
    
}

+ (NSString *)usernameInKeychain
{
    KeychainItemWrapper *keychainItem = [[KeychainItemWrapper alloc] initWithIdentifier:kKeychainId accessGroup:nil];
    NSString *username = [keychainItem objectForKey:(__bridge id)(kSecAttrAccount)];
    if ([username length] <= 0) username = nil;
    return username;
}

+ (void)resetKeychain
{
    KeychainItemWrapper *keychainItem = [[KeychainItemWrapper alloc] initWithIdentifier:kKeychainId accessGroup:nil];
    [keychainItem resetKeychainItem];
}

@end

File: ETSMobile/Controllers/Base/ETSAuthenticationViewController.swift
//
//  ETSAuthenticationViewController.swift
//  ETSMobile
//
//  Created by Thomas Durand on 13/05/2015.
//  Copyright (c) 2015 ApplETS. All rights reserved.
//

import UIKit

protocol ETSAuthenticationViewControllerDelegate {
    func controllerDidAuthenticate(controller: ETSAuthenticationViewController)
}

class ETSAuthenticationViewController: UIViewController, UITextFieldDelegate {
    
    let kKeychainId = "ApplETS"
    
    @IBOutlet weak var usernameTextField: UITextField!
    @IBOutlet weak var passwordTextField: UITextField!
    @IBOutlet weak var menuButton: UIButton!
    @IBOutlet weak var cancelButton: UIButton!
    @IBOutlet weak var loginButton: UIButton!
    
    var usernameInKeychain: String? {
        get {
            let keychainItem: KeychainItemWrapper = KeychainItemWrapper(idenfitier: "ApplETS", accessGroup: nil)
            var username: String = keychainItem.objectForKey(kSecAttrAccount);
            
            if count(username) <= 0 {
                return nil
            }
            
            return username;
        }
    }
    var passwordInKeychain: String? {
        get {
            let keychainItem: KeychainItemWrapper = KeychainItemWrapper(idenfitier: "ApplETS", accessGroup: nil)
            var password: String = keychainItem.objectForKey(kSecValueData);
            
            if count(password) <= 0 {
                return nil
            }
            
            return password;
        }
    }
    
    @IBAction func panLeftMenu(sender: AnyObject?) {
        let dynamicsDrawerViewController: MSDynamicsDrawerViewController = UIApplication.sharedApplication().dynamicsDrawerViewController
        dynamicsDrawerViewController.setPaneState(MSDynamicsDrawerPaneStateOpen, animated:true, allowUserInterruption:true, completion: nil)
    }
    
    func validateJSONResponse(response: [String: AnyObject]) -> ETSSynchronizationResponse {
        let apiError = response["d.erreur"]
        let requestError = response["Message"]
        
        if apiError is String {
            if apiError == "Code d'accès ou mot de passe invalide" {
                return ETSSynchronizationResponseAuthenticationError
            } else if count(apiError) > 0 {
                return ETSSynchronizationResponseUnknownError
            }
        } else if requestError is String {
            if requestError == "There was an error processing the request." {
                return ETSSynchronizationResponseAuthenticationError
            }
        }
        
        return ETSSynchronizationResponseValid;
    }
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        
        self.navigationController!.setNavigationBarHidden(UIDevice.currentDevice().userInterfaceIdiom == .Phone, animated: animated)
        
        if UIDevice.currentDevice().userInterfaceIdiom == .Pad {
            self.menuButton.hidden = true
            self.cancelButton.hidden = false
        } else {
            self.cancelButton.hidden = true
        }
        
        let dynamicsDrawerViewController: MSDynamicsDrawerViewController = UIApplication.sharedApplication().dynamicsDrawerViewController
        dynamicsDrawerViewController.setPaneDragRevealEnabled(true, forDirection: MSDynamicsDrawerDirectionLeft)
    }
    
    override func viewDidAppear(animated: Bool) {
        super.viewDidAppear(animated)
        self.usernameTextField.becomeFirstResponder()
    }
    
    func textFieldShouldReturn(textField: UITextField) -> Bool {
        if textField == self.usernameTextField {
            self.passwordTextField.becomeFirstResponder()
        } else {
            if count(self.usernameTextField.text) == 0 {
                self.usernameTextField.becomeFirstResponder()
            }
            else {
                self.authenticate(nil);
            }
        }
        
        return true
    }
    
    @IBAction func authenticate(sender: AnyObject?) {
        self.usernameTextField.resignFirstResponder()
        self.passwordTextField.resignFirstResponder()
        
        let keychainItem: KeychainItemWrapper = KeychainItemWrapper(idenfitier: "ApplETS", accessGroup: nil)
        keychainItem.setObject(self.passwordTextField.text, forKey: kSecValueData )
        keychainItem.setObject(self.usernameTextField.text, forKey: kSecAttrAccount )
        
        self.delegate.controllerDidAuthenticate(self)
    }
    
    @IBAction func cancel(sender: AnyObject?) {
        self.dismissViewControllerAnimated(true, completion: nil)
    }
    
    func resetKeychain() {
        let keychainItem: KeychainItemWrapper = KeychainItemWrapper(idenfitier: "ApplETS", accessGroup: nil)
        keychainItem.resetKeychainItem()
    }
}